dGame/dBehaviors/BehaviorContext.cpp

```diff
@@ -289,46 +290,102 @@ void BehaviorContext::Reset() {
 
-	std::vector<LWOOBJID> targets;
-
-	if (entity == nullptr) {
-		Game::logger->Log("BehaviorContext", "Invalid entity for (%llu)!", this->originator);
-
-		return targets;
-	}
+	// if we aren't targeting anything, then clear the targets vector
+	if (!targetSelf && !targetEnemy && !targetFriend && !targetTeam && ignoreFactionList.empty() && includeFactionList.empty()) {
+		targets.clear();
+	} else {
+		auto* caster = EntityManager::Instance()->GetEntity(this->caster);
 
-	if (!ignoreFaction && !includeFaction) {
-		for (auto entry : entity->GetTargetsInPhantom()) {
-			auto* instance = EntityManager::Instance()->GetEntity(entry);
+		// if the caster is not there, return empty targets list
+		if (!caster) {
+			Game::logger->LogDebug("BehaviorContext", "Invalid caster for (%llu)!", this->originator);
+			targets.clear();
+			return;
+		}
 
-			if (instance == nullptr) {
-				continue;
+		auto* casterDestroyableComponent = caster->GetComponent<DestroyableComponent>();
+		if (!casterDestroyableComponent) return;
+
+		auto index = targets.begin();
+		while (index != targets.end()) {
+			auto candidate = *index;
+			if (!candidate)index = targets.erase(index);
+			else if (candidate == caster){
+				if (!targetSelf) index = targets.erase(index);
+				else index++;
+			} else {
+				if (!CheckTargetingRequirements(candidate)) {
+					index = targets.erase(index);
+				} else {
+					// get factions to check against
+					// CheckTargetingRequirements checks for a destroyable component
+					// but we check again because bounds check are great!
+					auto candidateDestroyableComponent = candidate->GetComponent<DestroyableComponent>();
+					if (!candidateDestroyableComponent) {
+						index = targets.erase(index);
+						continue;
+					}
+					auto candidateFactions = candidateDestroyableComponent->GetFactionIDs();
+
+					if (candidateFactions.empty()){
+						index = targets.erase(index);
+					} else {
+						if (candidateDestroyableComponent->GetIsDead()) {
+							index = targets.erase(index);
+							continue;
+						}
+						if (!CheckFactionList(includeFactionList, candidateFactions)){
+							if (targetTeam){
+								auto* team = TeamManager::Instance()->GetTeam(this->caster);
+								if (team){
+									if(std::find(team->members.begin(), team->members.end(), candidate->GetObjectID()) != team->members.end()){
+										index++;
+										continue;
+									}
+								}
+							}
+							auto isEnemy = casterDestroyableComponent->IsEnemy(candidate);
+							if (!targetFriend && !isEnemy) index = targets.erase(index);
+							else if(!targetEnemy && isEnemy) index = targets.erase(index);
+							else if(CheckFactionList(ignoreFactionList, candidateFactions))	index = targets.erase(index);
+							else index++;
+							continue;
+						}
+						index++;
+					}
+				}
 			}
-
-			targets.push_back(entry);
+			continue;
 		}
 	}
```

Comment:
Is there any way we can reduce the nesting here?  This goes quite deep and I think we could probably make this look and read much easier

Reply:
So with this, I knew that was going to be a comment.
I personally like nesting like that. but yeah, it does make it not as easy to read.
The reason I did it like this was because this is the flow in the client and it helped me make sure I had the checks correct. I can try to re-write it more flat, but that will take a bit to do and verify

Reply:
Another reason to try to reduce nesting is to reduce the number of control paths there are.  There are quite a few spots where you do the same logic so reducing the number of control flow paths would help speed and also make it more clear whats going on